File: go-backend/models/field_registry.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
)

// ============================================================
// FIELD TYPE REGISTRY
// ============================================================

// FieldTypeRegistry - Master registry of all field types
type FieldTypeRegistry struct {
	ID          string `gorm:"primaryKey" json:"id"`     // 'text', 'email', 'repeater', etc.
	Category    string `gorm:"not null" json:"category"` // primitive, container, layout, special
	Label       string `gorm:"not null" json:"label"`
	Description string `json:"description"`
	Icon        string `json:"icon"`
	Color       string `json:"color"`

	// Schema definitions (JSON Schema format)
	InputSchema   datatypes.JSON `gorm:"type:jsonb;default:'{}'" json:"input_schema"`
	StorageSchema datatypes.JSON `gorm:"type:jsonb;default:'{}'" json:"storage_schema"`
	ConfigSchema  datatypes.JSON `gorm:"type:jsonb;default:'{}'" json:"config_schema"`

	// Behavior flags
	IsContainer  bool `gorm:"default:false" json:"is_container"`
	IsSearchable bool `gorm:"default:true" json:"is_searchable"`
	IsSortable   bool `gorm:"default:true" json:"is_sortable"`
	IsFilterable bool `gorm:"default:true" json:"is_filterable"`
	IsEditable   bool `gorm:"default:true" json:"is_editable"`
	SupportsPII  bool `gorm:"default:false" json:"supports_pii"`

	// Rendering hints
	TableRenderer  string `json:"table_renderer,omitempty"`
	FormRenderer   string `json:"form_renderer,omitempty"`
	ReviewRenderer string `json:"review_renderer,omitempty"`

	// AI Integration
	AISchema datatypes.JSON `gorm:"type:jsonb;default:'{}'" json:"ai_schema"`

	// Semantic type mapping
	DefaultSemanticType string `json:"default_semantic_type,omitempty"`

	// Edit tracking settings
	TrackChanges  bool `gorm:"default:true" json:"track_changes"`
	RequireReason bool `gorm:"default:false" json:"require_reason"`

	CreatedAt time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updated_at"`
}

func (FieldTypeRegistry) TableName() string {
	return "field_type_registry"
}

// ============================================================
// BATCH OPERATIONS
// ============================================================

// BatchOperation - Groups bulk changes together
type BatchOperation struct {
	ID          uuid.UUID  `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()" json:"id"`
	WorkspaceID uuid.UUID  `gorm:"type:uuid;not null;index" json:"workspace_id"`
	TableID     *uuid.UUID `gorm:"type:uuid;index" json:"table_id,omitempty"`

	// Operation details
	OperationType string `gorm:"not null" json:"operation_type"` // bulk_update, bulk_delete, import, etc.
	Description   string `json:"description"`

	// Scope
	AffectedRowCount   int      `gorm:"default:0" json:"affected_row_count"`
	AffectedFieldNames []string `gorm:"type:text[]" json:"affected_field_names"`

	// Status
	Status       string `gorm:"default:'completed'" json:"status"` // pending, in_progress, completed, failed, rolled_back
	ErrorMessage string `json:"error_message,omitempty"`

	// Rollback support
	CanRollback  bool       `gorm:"default:true" json:"can_rollback"`
	RolledBackAt *time.Time `json:"rolled_back_at,omitempty"`
	RolledBackBy *uuid.UUID `gorm:"type:uuid" json:"rolled_back_by,omitempty"`

	// Authorship
	CreatedBy   uuid.UUID  `gorm:"type:uuid;not null" json:"created_by"`
	CreatedAt   time.Time  `gorm:"autoCreateTime" json:"created_at"`
	CompletedAt *time.Time `json:"completed_at,omitempty"`
}

func (BatchOperation) TableName() string {
	return "batch_operations"
}

// ============================================================
// ROW VERSIONS (History)
// ============================================================

// RowVersion - Complete row snapshot for history
type RowVersion struct {
	ID      uuid.UUID `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()" json:"id"`
	RowID   uuid.UUID `gorm:"type:uuid;not null;index" json:"row_id"`
	TableID uuid.UUID `gorm:"type:uuid;not null;index" json:"table_id"`

	// Version info
	VersionNumber int `gorm:"not null" json:"version_number"`

	// Complete data snapshot (stored unredacted)
	Data     datatypes.JSON `gorm:"type:jsonb;not null" json:"data"`
	Metadata datatypes.JSON `gorm:"type:jsonb;default:'{}'" json:"metadata"`

	// Change context
	ChangeType    string `gorm:"not null" json:"change_type"` // create, update, restore, import, ai_edit, approval, bulk
	ChangeReason  string `json:"change_reason,omitempty"`     // User-provided reason
	ChangeSummary string `json:"change_summary,omitempty"`    // Auto-generated summary

	// Batch operation reference
	BatchOperationID *uuid.UUID `gorm:"type:uuid;index" json:"batch_operation_id,omitempty"`

	// Authorship
	ChangedBy *uuid.UUID `gorm:"type:uuid" json:"changed_by,omitempty"`
	ChangedAt time.Time  `gorm:"autoCreateTime" json:"changed_at"`

	// AI context
	AIAssisted     bool       `gorm:"default:false" json:"ai_assisted"`
	AIConfidence   *float64   `json:"ai_confidence,omitempty"`
	AISuggestionID *uuid.UUID `gorm:"type:uuid" json:"ai_suggestion_id,omitempty"`

	// Archive/Delete support
	IsArchived       bool       `gorm:"default:false" json:"is_archived"`
	ArchivedAt       *time.Time `json:"archived_at,omitempty"`
	ArchivedBy       *uuid.UUID `gorm:"type:uuid" json:"archived_by,omitempty"`
	ArchiveExpiresAt *time.Time `json:"archive_expires_at,omitempty"`

	CreatedAt time.Time `gorm:"autoCreateTime" json:"created_at"`
}

func (RowVersion) TableName() string {
	return "row_versions"
}

// ============================================================
// FIELD CHANGES (Granular Diffs)
// ============================================================

// FieldChange - Granular field-level change tracking
type FieldChange struct {
	ID           uuid.UUID  `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()" json:"id"`
	RowVersionID uuid.UUID  `gorm:"type:uuid;not null;index" json:"row_version_id"`
	RowID        uuid.UUID  `gorm:"type:uuid;not null;index" json:"row_id"`
	FieldID      *uuid.UUID `gorm:"type:uuid;index" json:"field_id,omitempty"`

	// Field identification (preserved if field is deleted)
	FieldName  string `gorm:"not null" json:"field_name"`
	FieldType  string `gorm:"not null" json:"field_type"`
	FieldLabel string `json:"field_label,omitempty"`

	// Change data (stored unredacted)
	OldValue datatypes.JSON `gorm:"type:jsonb" json:"old_value,omitempty"`
	NewValue datatypes.JSON `gorm:"type:jsonb" json:"new_value,omitempty"`

	// Change classification
	ChangeAction string `gorm:"not null" json:"change_action"` // add, update, remove, reorder

	// For container types, track nested path
	NestedPath []string `gorm:"type:text[]" json:"nested_path,omitempty"` // e.g., ['activities', '0', 'role']

	// AI analysis
	SimilarityScore    *float64 `json:"similarity_score,omitempty"`     // 0-1, 1=identical
	SemanticChangeType string   `json:"semantic_change_type,omitempty"` // typo_fix, correction, addition, major_revision

	// PII tracking
	ContainsPII bool `gorm:"default:false" json:"contains_pii"`

	CreatedAt time.Time `gorm:"autoCreateTime" json:"created_at"`
}

func (FieldChange) TableName() string {
	return "field_changes"
}

// ============================================================
// CHANGE APPROVALS
// ============================================================

// ChangeApproval - Approval workflow for edits
type ChangeApproval struct {
	ID      uuid.UUID `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()" json:"id"`
	RowID   uuid.UUID `gorm:"type:uuid;not null;index" json:"row_id"`
	TableID uuid.UUID `gorm:"type:uuid;not null;index" json:"table_id"`

	// Pending changes (not yet applied)
	PendingData    datatypes.JSON `gorm:"type:jsonb;not null" json:"pending_data"`
	PendingChanges datatypes.JSON `gorm:"type:jsonb;not null" json:"pending_changes"`
	ChangeReason   string         `json:"change_reason,omitempty"`

	// Approval context
	RequiresApprovalFrom string     `json:"requires_approval_from"` // table_owner, workspace_admin, specific_user, any_reviewer
	SpecificApproverID   *uuid.UUID `gorm:"type:uuid" json:"specific_approver_id,omitempty"`

	// Stage reference (if from workflow)
	StageID *uuid.UUID `gorm:"type:uuid" json:"stage_id,omitempty"`

	// Status
	Status string `gorm:"default:'pending'" json:"status"` // pending, approved, rejected, expired, cancelled

	// Requester
	RequestedBy uuid.UUID `gorm:"type:uuid;not null" json:"requested_by"`
	RequestedAt time.Time `gorm:"autoCreateTime" json:"requested_at"`

	// Approver
	ReviewedBy  *uuid.UUID `gorm:"type:uuid" json:"reviewed_by,omitempty"`
	ReviewedAt  *time.Time `json:"reviewed_at,omitempty"`
	ReviewNotes string     `json:"review_notes,omitempty"`

	// Expiration
	ExpiresAt *time.Time `json:"expires_at,omitempty"`

	CreatedAt time.Time `gorm:"autoCreateTime" json:"created_at"`
}

func (ChangeApproval) TableName() string {
	return "change_approvals"
}

// ============================================================
// AI FIELD SUGGESTIONS
// ============================================================

// AIFieldSuggestion - AI-powered field improvements
type AIFieldSuggestion struct {
	ID          uuid.UUID  `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()" json:"id"`
	WorkspaceID uuid.UUID  `gorm:"type:uuid;not null;index" json:"workspace_id"`
	TableID     uuid.UUID  `gorm:"type:uuid;not null;index" json:"table_id"`
	RowID       *uuid.UUID `gorm:"type:uuid;index" json:"row_id,omitempty"` // NULL for table-level suggestions
	FieldID     *uuid.UUID `gorm:"type:uuid;index" json:"field_id,omitempty"`

	// Suggestion type
	SuggestionType string `gorm:"not null" json:"suggestion_type"` // typo_correction, format_correction, etc.

	// Current state
	CurrentValue datatypes.JSON `gorm:"type:jsonb" json:"current_value,omitempty"`

	// Suggestion
	SuggestedValue datatypes.JSON `gorm:"type:jsonb;not null" json:"suggested_value"`
	Confidence     float64        `gorm:"not null" json:"confidence"` // 0-1
	Reasoning      string         `json:"reasoning,omitempty"`

	// Evidence
	SampleData     datatypes.JSON `gorm:"type:jsonb" json:"sample_data,omitempty"`
	PatternMatches *int           `json:"pattern_matches,omitempty"`
	TotalValues    *int           `json:"total_values,omitempty"`

	// Related suggestions
	RelatedSuggestionIDs []uuid.UUID `gorm:"type:uuid[]" json:"related_suggestion_ids,omitempty"`

	// Status
	Status string `gorm:"default:'pending'" json:"status"` // pending, accepted, rejected, dismissed, auto_applied

	// Review
	ReviewedBy  *uuid.UUID `gorm:"type:uuid" json:"reviewed_by,omitempty"`
	ReviewedAt  *time.Time `json:"reviewed_at,omitempty"`
	ReviewNotes string     `json:"review_notes,omitempty"`

	// If applied
	AppliedVersionID *uuid.UUID `gorm:"type:uuid" json:"applied_version_id,omitempty"`

	CreatedAt time.Time  `gorm:"autoCreateTime" json:"created_at"`
	ExpiresAt *time.Time `json:"expires_at,omitempty"`
}

func (AIFieldSuggestion) TableName() string {
	return "ai_field_suggestions"
}

// ============================================================
// CONSTANTS
// ============================================================

// Field type categories
const (
	FieldCategoryPrimitive = "primitive"
	FieldCategoryContainer = "container"
	FieldCategoryLayout    = "layout"
	FieldCategorySpecial   = "special"
)

// Change types
const (
	ChangeTypeCreate   = "create"
	ChangeTypeUpdate   = "update"
	ChangeTypeRestore  = "restore"
	ChangeTypeImport   = "import"
	ChangeTypeAIEdit   = "ai_edit"
	ChangeTypeApproval = "approval"
	ChangeTypeBulk     = "bulk"
)

// Change actions
const (
	ChangeActionAdd     = "add"
	ChangeActionUpdate  = "update"
	ChangeActionRemove  = "remove"
	ChangeActionReorder = "reorder"
)

// Approval statuses
const (
	ApprovalStatusPending   = "pending"
	ApprovalStatusApproved  = "approved"
	ApprovalStatusRejected  = "rejected"
	ApprovalStatusExpired   = "expired"
	ApprovalStatusCancelled = "cancelled"
)

// Suggestion types
const (
	SuggestionTypeTypoCorrection     = "typo_correction"
	SuggestionTypeFormatCorrection   = "format_correction"
	SuggestionTypeSemanticTypeChange = "semantic_type_change"
	SuggestionTypeValidationRule     = "validation_rule"
	SuggestionTypeNormalizeValues    = "normalize_values"
	SuggestionTypeMissingValue       = "missing_value"
	SuggestionTypeDuplicateDetection = "duplicate_detection"
)

// Suggestion statuses
const (
	SuggestionStatusPending     = "pending"
	SuggestionStatusAccepted    = "accepted"
	SuggestionStatusRejected    = "rejected"
	SuggestionStatusDismissed   = "dismissed"
	SuggestionStatusAutoApplied = "auto_applied"
)

File: src/types/field-registry.ts
/**
 * Field Registry Types
 * Universal field type definitions for consistent data handling
 * 
 * NOTE: Core field types are now defined in field-types.ts
 * This file re-exports those types and adds history/version types
 */

// ============================================================
// RE-EXPORTS FROM FIELD-TYPES.TS (NEW CANONICAL SOURCE)
// ============================================================

// Import for local use
import type { FieldCategory as FC } from './field-types';

export type { 
  FieldCategory, 
  FieldTypeRegistry, 
  AISchema,
  Field,
  EffectiveFieldConfig,
  FieldTypeSummary,
  FieldTypesByCategory,
} from './field-types';

export { 
  FIELD_TYPES, 
  isContainerField,
  getFieldTypeId,
} from './field-types';

// Legacy type alias for backwards compatibility
export type FieldTypeId =
  | 'text' | 'textarea' | 'email' | 'phone' | 'url' | 'address'
  | 'number' | 'rating'
  | 'date' | 'datetime' | 'time'
  | 'select' | 'multiselect' | 'radio' | 'checkbox'
  | 'group' | 'repeater' | 'section'
  | 'divider' | 'heading' | 'paragraph' | 'callout'
  | 'file' | 'image' | 'signature'
  | 'rank' | 'item_list'

// ============================================================
// ROW VERSIONS (History)
// ============================================================

export type ChangeType = 'create' | 'update' | 'restore' | 'import' | 'ai_edit' | 'approval' | 'bulk'
export type ChangeAction = 'add' | 'update' | 'remove' | 'reorder'

export interface RowVersion {
  id: string
  row_id: string
  table_id: string
  
  version_number: number
  
  // Data snapshot (may be redacted if PII mode enabled)
  data: Record<string, any>
  metadata: Record<string, any>
  
  // Change context
  change_type: ChangeType
  change_reason?: string
  change_summary?: string
  
  // Batch reference
  batch_operation_id?: string
  
  // Authorship
  changed_by?: string
  changed_at: string
  
  // AI context
  ai_assisted: boolean
  ai_confidence?: number
  ai_suggestion_id?: string
  
  // Archive status
  is_archived: boolean
  archived_at?: string
  archived_by?: string
  archive_expires_at?: string
  
  created_at: string
}

export interface FieldChange {
  id: string
  row_version_id: string
  row_id: string
  field_id?: string
  
  field_name: string
  field_type: string
  field_label?: string
  
  old_value?: any
  new_value?: any
  
  change_action: ChangeAction
  nested_path?: string[]  // For repeater/group: ['activities', '0', 'role']
  
  similarity_score?: number  // 0-1, 1 = identical
  semantic_change_type?: 'typo_fix' | 'correction' | 'addition' | 'major_revision'
  
  contains_pii: boolean
  created_at: string
}

export interface RowHistoryEntry extends Omit<RowVersion, 'data'> {
  data: Record<string, any>  // May be redacted
  field_changes?: FieldChange[]
}

// ============================================================
// BATCH OPERATIONS
// ============================================================

export type BatchOperationType = 'bulk_update' | 'bulk_delete' | 'bulk_create' | 'import' | 'ai_correction' | 'restore'
export type BatchOperationStatus = 'pending' | 'in_progress' | 'completed' | 'failed' | 'rolled_back'

export interface BatchOperation {
  id: string
  workspace_id: string
  table_id?: string
  
  operation_type: BatchOperationType
  description?: string
  
  affected_row_count: number
  affected_field_names?: string[]
  
  status: BatchOperationStatus
  error_message?: string
  
  can_rollback: boolean
  rolled_back_at?: string
  rolled_back_by?: string
  
  created_by: string
  created_at: string
  completed_at?: string
}

// ============================================================
// CHANGE APPROVALS
// ============================================================

export type ApprovalRequirement = 'table_owner' | 'workspace_admin' | 'specific_user' | 'any_reviewer'
export type ApprovalStatus = 'pending' | 'approved' | 'rejected' | 'expired' | 'cancelled'

export interface ChangeApproval {
  id: string
  row_id: string
  table_id: string
  
  pending_data: Record<string, any>
  pending_changes: FieldChange[]
  change_reason?: string
  
  requires_approval_from: ApprovalRequirement
  specific_approver_id?: string
  
  stage_id?: string
  
  status: ApprovalStatus
  
  requested_by: string
  requested_at: string
  
  reviewed_by?: string
  reviewed_at?: string
  review_notes?: string
  
  expires_at?: string
  created_at: string
}

// ============================================================
// AI FIELD SUGGESTIONS
// ============================================================

export type SuggestionType =
  | 'typo_correction'
  | 'format_correction'
  | 'semantic_type_change'
  | 'validation_rule'
  | 'field_type_change'
  | 'merge_fields'
  | 'split_field'
  | 'normalize_values'
  | 'missing_value'
  | 'duplicate_detection'

export type SuggestionStatus = 'pending' | 'accepted' | 'rejected' | 'dismissed' | 'auto_applied'

export interface AIFieldSuggestion {
  id: string
  workspace_id: string
  table_id: string
  row_id?: string
  field_id?: string
  
  suggestion_type: SuggestionType
  
  current_value?: any
  suggested_value: any
  confidence: number  // 0-1
  reasoning?: string
  
  sample_data?: any
  pattern_matches?: number
  total_values?: number
  
  related_suggestion_ids?: string[]
  
  status: SuggestionStatus
  
  reviewed_by?: string
  reviewed_at?: string
  review_notes?: string
  
  applied_version_id?: string
  
  created_at: string
  expires_at?: string
}

// ============================================================
// TABLE SETTINGS
// ============================================================

export interface TableHistorySettings {
  track_changes: boolean
  require_change_reason: boolean
  version_retention_days: number | null  // null = keep forever
  allow_user_delete_history: boolean
  admin_only_full_history: boolean
}

export interface TableApprovalSettings {
  require_approval: boolean
  approval_type: ApprovalRequirement
  specific_approvers: string[]  // User IDs
  auto_expire_days: number
  notify_on_pending: boolean
}

export interface TableAISettings {
  enable_suggestions: boolean
  auto_apply_high_confidence: boolean
  auto_apply_threshold: number  // e.g., 0.95
  suggestion_types: SuggestionType[]
}

// ============================================================
// API TYPES
// ============================================================

// GET /api/v1/tables/:id/rows/:row_id/history
export interface GetRowHistoryRequest {
  redact_pii?: boolean
  include_archived?: boolean
  limit?: number
}

export interface GetRowHistoryResponse {
  row_id: string
  total_versions: number
  versions: RowHistoryEntry[]
}

// POST /api/v1/tables/:id/rows/:row_id/restore/:version
export interface RestoreVersionRequest {
  reason: string
}

export interface RestoreVersionResponse {
  success: boolean
  new_version_id: string
  new_version_number: number
}

// GET /api/v1/tables/:id/rows/:row_id/diff/:v1/:v2
export interface VersionDiffResponse {
  version1: number
  version2: number
  field_diffs: FieldChange[]
}

// POST /api/v1/tables/:id/rows/:row_id/approve
export interface ApproveChangeRequest {
  approval_id: string
  action: 'approve' | 'reject'
  notes?: string
}

// GET /api/v1/tables/:id/ai/suggestions
export interface GetAISuggestionsRequest {
  status?: SuggestionStatus
  suggestion_type?: SuggestionType
  min_confidence?: number
  limit?: number
}

export interface GetAISuggestionsResponse {
  table_id: string
  suggestions: AIFieldSuggestion[]
  total: number
}

// POST /api/v1/tables/:id/ai/suggestions/:id/apply
export interface ApplySuggestionRequest {
  suggestion_id: string
  apply: boolean
  notes?: string
}

// ============================================================
// HELPER TYPES
// ============================================================

// Type-safe field value based on field type
export type FieldValue<T extends FieldTypeId> =
  T extends 'repeater' ? Array<Record<string, any>> :
  T extends 'group' ? Record<string, any> :
  T extends 'multiselect' | 'rank' | 'item_list' ? string[] :
  T extends 'checkbox' ? boolean :
  T extends 'number' | 'rating' ? number :
  T extends 'file' | 'image' ? FileValue :
  T extends 'divider' | 'heading' | 'paragraph' | 'callout' | 'section' ? null :
  string

export interface FileValue {
  url: string
  name: string
  size?: number
  mime_type?: string
  width?: number  // For images
  height?: number // For images
}

// Container field with children
export interface ContainerFieldConfig {
  children: Array<{
    name: string
    label: string
    type: FieldTypeId
    is_required?: boolean
    placeholder?: string
    options?: string[]
    validation?: Record<string, any>
  }>
  min_items?: number  // For repeaters
  max_items?: number  // For repeaters
  item_label?: string // For repeaters
}

// ============================================================
// CONSTANTS
// ============================================================

export const FIELD_CATEGORIES: Record<FC, FieldTypeId[]> = {
  primitive: ['text', 'textarea', 'email', 'phone', 'url', 'number', 'date', 'datetime', 'time', 'select', 'multiselect', 'radio', 'checkbox'],
  container: ['group', 'repeater', 'section'],
  layout: ['divider', 'heading', 'paragraph', 'callout'],
  special: ['file', 'image', 'signature', 'rating', 'rank', 'item_list'],
}

export const PII_FIELD_TYPES: FieldTypeId[] = ['email', 'phone', 'signature', 'address']

export const CONTAINER_FIELD_TYPES: FieldTypeId[] = ['group', 'repeater', 'section']

export const LAYOUT_FIELD_TYPES: FieldTypeId[] = ['divider', 'heading', 'paragraph', 'callout']
